### ISSUE TITLE
New `flatten` transform

### ISSUE THREAD

**Issue by binarylogic** [timberio/vector#1880]
This transform would recursively flatten all nested keys into the root document. For example

## Example 1: Default root key

```toml
[transforms.flatten_all]
  type = "flatten"
```

Input:

```json
{
  "parent": {
    "child": {
      "grandchild": "value"
    }
  }
}
```

Output:

```json
{
  "grandchild": "value"
}
```

## Example 2: Configurable root key

```toml
[transforms.flatten_all]
  type = "flatten"
  root = "parent"
```

Input:

```json
{
  "parent": {
    "child": {
      "grandchild": "value"
    }
  }
}
```

Output:

```json
{
  "parent": {
    "grandchild": "value"
  }
}
```

**Comment by a-rodin** [comment]
What should be the behavior in case if the input looks like this:

```json
{
  "parent": {
    "child1": {
      "grandchild": "value1"
    },
    "child2": {
      "grandchild": "value2"
    }
  }
}
```

Note that as generally Vector changes order of the fields (both now and after with #1838), it is not possible to use "use the first value" or "use the last value" semantics. But it will be possible to use "use the first value in the alphabetic order" or "use the last value in the alphabetic order" after #1838.

**Comment by binarylogic** [comment]
Good question, you have a knack for interesting edge-cases. I would expect the last value to win (`value2`). So:

```json
{
  "grandchild": "value2"
}
```

But I don't think this should be a consistent requirement since this edge-case will not be common. So, we shouldn't make any guarantee on ordering or priority on values. If it happens that we are consistent, then that's better.

**Comment by Alexx-G** [comment]
Also it might be useful to support `level`. I imagine there might be use cases when you want to flatten the structure but stopping at a certain level.

**Comment by binarylogic** [comment]
Closing in favor of https://github.com/timberio/vector/issues/3752
